src/contracts/MultiLP.sol

```diff
@@ -0,0 +1,231 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.23;
+
+import {IERC20, ERC20Upgradeable} from "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
+import {SafeCast} from "@openzeppelin/contracts/utils/math/SafeCast.sol";
+
+import {AbstractARM} from "./AbstractARM.sol";
+
+/**
+ * @title Abstract support to an ARM for multiple Liquidity Providers (LP)
+ * @author Origin Protocol Inc
+ */
+abstract contract MultiLP is AbstractARM, ERC20Upgradeable {
+    /// @notice The delay before a withdrawal request can be claimed in seconds
+    uint256 public constant CLAIM_DELAY = 10 minutes;
+    /// @dev The amount of shares that are minted to a dead address on initalization
+    uint256 internal constant MIN_TOTAL_SUPPLY = 1e12;
+    /// @dev The address with no know private key that the initial shares are minted to
+    address internal constant DEAD_ACCOUNT = 0x000000000000000000000000000000000000dEaD;
+
+    /// @notice The address of the liquidity token. eg WETH
+    address internal immutable liquidityToken;
+
+    struct WithdrawalQueueMetadata {
+        // cumulative total of all withdrawal requests included the ones that have already been claimed
+        uint128 queued;
+        // cumulative total of all the requests that can be claimed including the ones that have already been claimed
+        uint128 claimable;
+        // total of all the requests that have been claimed
+        uint128 claimed;
+        // index of the next withdrawal request starting at 0
+        uint128 nextWithdrawalIndex;
+    }
+
+    /// @notice Global metadata for the withdrawal queue including:
+    /// queued - cumulative total of all withdrawal requests included the ones that have already been claimed
+    /// claimable - cumulative total of all the requests that can be claimed including the ones already claimed
+    /// claimed - total of all the requests that have been claimed
+    /// nextWithdrawalIndex - index of the next withdrawal request starting at 0
+    // slither-disable-next-line uninitialized-state
+    WithdrawalQueueMetadata public withdrawalQueueMetadata;
+
+    struct WithdrawalRequest {
+        address withdrawer;
+        bool claimed;
+        uint40 timestamp; // timestamp of the withdrawal request
```

Comment:
Let's turn this around and store (and emit) the time that the withdraw request unlocks. This makes for easier contract upgrades if we ever decide to change the withdraw time, and also means that the front-end / analytics knows when these unlock, even without talking to the contract, or syncing "constants".

Reply:
Good idea. I've changed and pushed